compiler/llvm/llvmUtil.cpp

```diff
@@ -496,8 +496,14 @@ llvm::Value *convertValueToType(llvm::IRBuilder<>* irBuilder,
 
   //Pointers
   if(newType->isPointerTy() && curType->isPointerTy()) {
-    if( newType->getPointerAddressSpace() !=
-        curType->getPointerAddressSpace() ) {
+
+    // It's safe to convert pointers between generic (0) and non generic (non
+    // 0) address spaces
+    auto newTypeAddrSpace = newType->getPointerAddressSpace();
+    auto curTypeAddrSpace = curType->getPointerAddressSpace();
+    if(newTypeAddrSpace != 0 && curTypeAddrSpace != 0 && newTypeAddrSpace !=
+      curTypeAddrSpace)
```

Comment:
pedantic: break line after the last `&&`, align `newTypeAddrSpace`.

not-so-pedantic: Your comment/change makes sense to me, but I don't trust my judgement here at all. Do you have an explanation as to why this hasn't been an issue so far, but it is now?

Reply:
Michael would be the person to ask, if you are unsure about this.

Reply:
I go into some more detail about it in this comment here: https://github.com/Cray/chapel-private/issues/4403#issuecomment-1447329159

My guess as to why it wasn't an issue before is that the AMD backend is being more specific about what address space local data (data allocated within a GPU kernel) goes into than the NVPTX backend is.

Reply:
I see, thanks. Generality of this worries me a little bit. A full paratest with non-gpu locale model using LLVM backend can be a good idea.

Reply:
> Generality of this worries me a little bit

That's understandable. What's giving me some more confidence is the IR getting produced by the call to `CreatePointerCast` is adding in the needed `addrspacecast` at least for this specific example I'm working with: 

`%58 = addrspacecast [1 x i64] addrspace(5)* %21 to [1 x i64]*`

I would also hope `irBuilder->CreatePointerCast()` itself would trigger an assertion if we were doing something nonsensical.

FWIW I've run this through paratests and the flat locale model and it looks fine though I'm not sure how revealing that is (I don't imagine we have any tests that would lock down this assertion).